clients/TypeScript/packages/client/src/ChainSync/ChainSyncClient.ts

```diff
@@ -6,90 +6,85 @@ import {
   createClientContext
 } from '../Connection'
 import { UnknownResultError } from '../errors'
-import { baseRequest } from '../Request'
 import { createPointFromCurrentTip, ensureSocketIsOpen } from '../util'
 import { findIntersect, Intersection } from './findIntersect'
+import { requestNext } from './requestNext'
 
 export interface ChainSyncClient {
   context: InteractionContext
-  findIntersect: (points: Point[]) => ReturnType<typeof findIntersect>
-  initialIntersection: Intersection
-  on: (messageHandlers: {
-    rollBackward: (response: {
-      point: Point,
-      tip: Tip,
-      reflection: Mirror
-    }) => void
-    rollForward: (response: {
-      block: Block,
-      tip: Tip,
-      reflection: Mirror
-    }) => void
-  }) => void
   requestNext: (options?: { mirror?: Mirror }) => void
   shutdown: () => Promise<void>
+  startSync: (points?: Point[], requestBuffer?: number) => Promise<Intersection>
+}
+
+export interface ChainSyncMessageHandlers {
+  rollBackward: (response: {
+    point: Point,
+    tip: Tip,
+    reflection: Mirror
+  }) => void
+  rollForward: (response: {
+    block: Block,
+    tip: Tip,
+    reflection: Mirror
+  }) => void
 }
 
-export const createChainSyncClient = async (options?: {
+export const createChainSyncClient = async (
+  messageHandlers: ChainSyncMessageHandlers,
+  options?: {
   connection?: ConnectionConfig
 }): Promise<ChainSyncClient> => {
   return new Promise((resolve, reject) => {
     createClientContext(options).then(context => {
       const { socket } = context
       socket.once('error', reject)
-      socket.once('open', async () => {
-        const initialIntersection = await findIntersect(
-          [await createPointFromCurrentTip({ socket, closeOnCompletion: false })],
-          {
-            socket,
-            closeOnCompletion: false
+      socket.on('message', (message: string) => {
+        const response: Ogmios['RequestNextResponse'] = JSON.parse(message)
+        if (response.methodname === 'RequestNext') {
+          if ('RollBackward' in response.result) {
+            messageHandlers.rollBackward({
+              point: response.result.RollBackward.point,
+              tip: response.result.RollBackward.tip,
+              reflection: response.reflection
+            })
+          } else if ('RollForward' in response.result) {
+            messageHandlers.rollForward({
+              block: response.result.RollForward.block,
+              tip: response.result.RollForward.tip,
+              reflection: response.reflection
+            })
+          } else {
+            throw new UnknownResultError(response.result)
```

Comment:
I am still a bit doubtful about this w.r.t. concurrency. I assume this is fine if a new connection is used for the state queries, but if re-used, then this may cause issues? I should check back how the state query client is implemented maybe to dispel my doubts.

Reply:
Each client has an isolated connection, however, the scenario you're describing is possible by passing in an `InteractionContext` so is worth documenting with a test.

Reply:
https://github.com/KtorZ/cardano-ogmios/issues/40